### ISSUE TITLE
Balancing IoUD and IoUI/C

### ISSUE THREAD

**Issue by haruishi43** [zifuwanggg/JDTLosses#3]
Thank you for your work.
I noticed that in a distributed training setting where the batch per GPU is low, the JML will lean towards optimizing IoU/C and will have lower overall IoUD. I would like to mitigate this so that the training will balance IoU and IoU/C better. Do you have any tips regarding this? Would changing the loss weights between CE and JML help? I currently use a 0.25:0.75 split as you recommended.

**Comment by zifuwanggg** [comment]
Thank you for your interest in our work.

Firstly, I believe that adjusting the loss weights would likely reduce the impact of JML, which could result in overall worse performance, rather than just improving IoUD.

Additionally, the current loss implementation does not support multi-GPU gathering. However, you might find the implementation from [nnUNet](https://github.com/MIC-DKFZ/nnUNet/blob/3df680c6d28ea88735fca1c1921df380471c1af6/nnunetv2/training/loss/dice.py#L35) useful as a potential solution.

**Comment by haruishi43** [comment]
I see, thanks for the helpful tip.
I was just using your JML on mmseg using DDP and was getting lower IoU scores (around 2% lower) on a custom remote-sensing dataset and didn't realize your original codebase didn't support multi-GPU.
I might try to customize your JML with the nnUNet as you recommended to see if it improves.

**Comment by haruishi43** [comment]
@zifuwanggg I updated the loss function by following `nnUNet`'s `AllGather` function to gather `tp, fp, fn` and `active_classes` before calculating the losses in JDT. I noticed that IoU is significantly better than before! However, it is still a little low. Do you think there could be further improvements to further improve the IoU using your JML?

For reference, this is the IoU/IoUI/IoUC:
![Image](https://github.com/user-attachments/assets/a2e0f407-32c0-4712-b111-6dbea67c10ae)
- Red: CE
- Blue: JML (non-DDP)
- Green: JML (DDP)

**Comment by zifuwanggg** [comment]
@haruishi43 This is a very interesting result! Have you had a chance to check the [recommendations](https://github.com/zifuwanggg/JDTLosses/blob/master/FAQs.md)? One thing you could do, if you haven't already, is to reduce the number of iterations. Additionally, if your dataset contains many classes, it's important to make sure that you calculate the loss based only on the classes present in the mini-batch, rather than using all classes.

**Comment by haruishi43** [comment]
I checked out the recommendations and read your paper. Since the number of classes is similar to Cityscapes, I’m using your default configurations and kept `active_classes_mode_hard=“present”`. I’ll keep looking into this.

**Comment by haruishi43** [comment]
Hey @zifuwanggg, I think I might have made a mistake with the CE part of the JML loss. I’ve been using the traditional cross-entropy loss without label smoothing instead of your KLDiv-based “soft” cross-entropy loss. In your `get_criterion()` function, I noticed that you used FocalLoss for hard labels and KLDiv when using label smoothing. I actually used the combination of JML-BLS with OHEM Cross-Entropy Loss (non-LS). I guess this might be one of the reasons why the performance was weird. However, I have a concern about using KLDiv as CE since I noticed that normal CE is worse than OHEM CE for my dataset, so I wonder if using KLDiv would even improve IoUD.

**Comment by zifuwanggg** [comment]
KLDiv and CE differ only by a constant and have the same gradient during optimization. The reason I use KLDiv is that PyTorch's CE doesn't support BLS (it has to smooth all pixels).

While I don't have extensive experience with OHEM, the core idea seems to be selecting hard examples during training. From this perspective, you should first identify the hard pixels, then apply both CE and JML to those selected pixels. It looks like you're only applying CE to the selected pixels, but JML to all pixels, which could lead to issues.

**Comment by haruishi43** [comment]
@zifuwanggg, thanks for the feedback. I guess the natural way to go about this is to apply OHEM to your JML so it can focus on the more difficult pixels.
